src/helics/core/BasicHandleInfo.hpp

```diff
@@ -56,14 +58,25 @@ class BasicHandleInfo {
     const std::string key;  //!< the name of the handle
     const std::string type;  //!< the type of data used by the handle
     const std::string units;  //!< the units associated with the handle
-    std::string interface_info;  //!< storage for a user info string
     const std::string& type_in;  //!< the input type of a filter
     const std::string& type_out;  //!< the output type of a filter
     /** get the interface handle information */
     InterfaceHandle getInterfaceHandle() const { return handle.handle; }
     /** extract a global federate id */
     GlobalFederateId getFederateId() const { return handle.fed_id; }
     /** set the user level information field*/
-    void setInfoField(std::string& info) { interface_info = info; }
+    void setTag(const std::string& tag, const std::string& value);
+    /** search for a tag by name*/
+    const std::string& getTag(const std::string& tag) const;
+    /** get a tag (key, value) pair by index*/
+    const std::pair<std::string, std::string>& getTagByIndex(size_t index) const
+    {
+        return tags[index];
+    }
+    /** get the number of tags associated with an interface*/
+    auto tagCount() const { return tags.size(); }
+
+  private:
+    std::vector<std::pair<std::string, std::string>> tags;  //!< storage for a user info string
```

Comment:
I started with std::map,  could move back to it, but the likely use case is a fairly low number of tags 0 being common, maybe 4 in a couple cases.    so having a map in the object seemed a bit heavy for what it was doing.    One issue with making the json map is the need to loop over them to see what tags a particular interface has.  And there is no restriction on having json compatible field names.    So a vector seemed a bit more natural for the operations that are needed internally and is likely to be faster though speed is not likely to be a big issue here.

Reply:
The empty map and per item overheads seem like a good reason to keep using an std::vector.

If we're using a json parser that follows the standard then the restrictions on json compatible field names are identical to valid string values; any unicode codepoint except `"` or `\` or control characters (which must be escaped). It looks like toml quoted keys are similar.

For the serialized version, serializing it to a json map (jsoncpp `getMemberNames()`) and possibly an inline toml table (toml11 key-value iterator/foreach) would make it nicer for users editing the config files (less brackets and less verbose), and using query responses in other languages (i.e. javascript `response["tags"]["tagname"]` vs writing a custom array search function). The test files would become:

json:
```json
"tags": {
    "period":  "0.5" ,
    "description": "a test publication"
}
```

toml:
```toml
tags = {"period"="0.5", "description"="a test publication"}
```

Reply:
Ok I like that for formatting purposes somewhat better.  I think I still want to support the name value keys since then you specify things in a json schema if desired.  But supporting the general format as described above is probably better.  This would require putting some limitations on tag names that would need to be checked in the core and would generate an error if a tag string was not json compatible.

Reply:
Actually, here are some interesting things in jsoncpp:
- https://github.com/open-source-parsers/jsoncpp/blob/master/src/lib_json/json_writer.cpp#L280
- https://github.com/open-source-parsers/jsoncpp/blob/master/src/lib_json/json_writer.cpp#L493

It looks like the `valueToQuotedStringN` function is called by `valueToQuotedString`, which is used when writing string values and member names, so we might not need to do our own checking.